File: array/TernarySearch/java/TernarySearch.java
public final class TernarySearch {
    private TernarySearch() {}

    public static final class Result {
        private final int index;
        private final long value;

        public Result(int index, long value) {
            this.index = index;
            this.value = value;
        }

        public int index() {
            return index;
        }

        public long value() {
            return value;
        }
    }

    public static Result findUnimodalMaximum(long[] values) {
        if (values.length == 0) {
            throw new IllegalArgumentException("ternary search requires a non-empty array");
        }

        int left = 0;
        int right = values.length - 1;

        while (right > left + 3) {
            int third = (right - left) / 3;
            int mid1 = left + third;
            int mid2 = right - third;

            if (values[mid1] < values[mid2]) {
                left = mid1 + 1;
            } else if (values[mid1] > values[mid2]) {
                right = mid2 - 1;
            } else {
                left = mid1;
                right = mid2;
            }
        }

        int bestIndex = left;
        for (int index = left + 1; index <= right; index++) {
            if (values[index] > values[bestIndex]) {
                bestIndex = index;
            }
        }

        return new Result(bestIndex, values[bestIndex]);
    }
}

File: array/TernarySearch/rust/src/ternary_search.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TernarySearchResult {
    pub index: usize,
    pub value: i64,
}

pub fn find_unimodal_maximum(values: &[i64]) -> Result<TernarySearchResult, String> {
    if values.is_empty() {
        return Err(String::from("ternary search requires a non-empty array"));
    }

    let mut left = 0usize;
    let mut right = values.len() - 1;

    while right > left + 3 {
        let third = (right - left) / 3;
        let mid1 = left + third;
        let mid2 = right - third;

        if values[mid1] < values[mid2] {
            left = mid1 + 1;
        } else if values[mid1] > values[mid2] {
            right = mid2 - 1;
        } else {
            left = mid1;
            right = mid2;
        }
    }

    let mut best_index = left;
    for index in (left + 1)..=right {
        if values[index] > values[best_index] {
            best_index = index;
        }
    }

    Ok(TernarySearchResult {
        index: best_index,
        value: values[best_index],
    })
}
